File: apps/ai-service/app/schemas/review.py
from typing import Literal, Optional
from pydantic import BaseModel, Field


CodeType = Literal[
    "react-component",
    "express-route",
    "express-controller",
    "mongoose-model",
    "javascript-utility",
]

ReviewMode = Literal[
    "quick-review",
    "deep-review",
    "security-focused",
    "performance-focused",
    "learning-mode",
]

Severity = Literal["Critical", "High", "Medium", "Low", "Suggestion"]

IssueCategory = Literal[
    "Bug",
    "Security",
    "Performance",
    "Best Practice",
    "Maintainability",
    "Accessibility",
]


class AnalyzeReviewRequest(BaseModel):
    fileName: str = Field(..., min_length=1, max_length=120)
    codeType: CodeType
    reviewMode: ReviewMode
    code: str = Field(..., min_length=10, max_length=50000)


class ReviewIssue(BaseModel):
    title: str = Field(..., description="Short title for the issue.")
    category: IssueCategory = Field(..., description="Main category of the issue.")
    severity: Severity = Field(..., description="Severity level of the issue.")
    lineNumber: Optional[int] = Field(
        default=None,
        description="Line number if confidently known, otherwise null.",
    )
    explanation: str = Field(
        ...,
        description="Clear explanation of why this is a problem.",
    )
    recommendation: str = Field(
        ...,
        description="Specific recommendation to fix the issue.",
    )
    codeExample: str = Field(
        default="",
        description="Optional small code example showing the fix.",
    )


class AnalyzeReviewResponse(BaseModel):
    score: int = Field(
        ...,
        ge=0,
        le=100,
        description="Overall code quality score from 0 to 100.",
    )
    summary: str = Field(
        ...,
        description="High-level summary of the code review.",
    )
    issues: list[ReviewIssue] = Field(
        ...,
        description="List of detected issues.",
    )
    improvedCode: str = Field(
        ...,
        description="Improved/refactored version of the submitted code.",
    )
    learningNotes: list[str] = Field(
        ...,
        description="Important learning points for the developer.",
    )
    recommendedTopics: list[str] = Field(
        ...,
        description="Topics the developer should study next.",
    )
    aiProvider: str = Field(
        default="gemini",
        description="AI provider/model used for the review.",
    )

File: apps/web/models/Review.ts
import mongoose, { Schema, type Model, Types } from "mongoose";

export interface IIssue {
  title: string;
  category: string;
  severity: "Critical" | "High" | "Medium" | "Low" | "Suggestion";
  severity_level?: "critical" | "high" | "medium" | "low";
  lineNumber?: number | null;
  explanation: string;
  recommendation: string;
  codeExample?: string;
}

export interface IReview {
  userId: Types.ObjectId;
  title: string;
  fileName: string;
  codeType: string;
  reviewMode: string;
  status: "pending" | "completed" | "failed" | "pending_approval" | "approved" | "rejected";
  originalCode: string;
  improvedCode?: string;
  score?: number | null;
  summary?: string;
  issues: IIssue[];
  learningNotes: string[];
  recommendedTopics: string[];
  aiProvider?: string;
  errorMessage?: string;
  threadId?: string;
  createdAt?: Date;
  updatedAt?: Date;
}

const issueSchema = new Schema<IIssue>(
  {
    title: { type: String, required: true },
    category: { type: String, required: true },
    severity: {
      type: String,
      enum: ["Critical", "High", "Medium", "Low", "Suggestion"],
      default: "Medium",
    },
    severity_level: {
      type: String,
      enum: ["critical", "high", "medium", "low"],
      default: "medium",
    },
    lineNumber: { type: Number, default: null },
    explanation: { type: String, required: true },
    recommendation: { type: String, required: true },
    codeExample: { type: String, default: "" },
  },
  { _id: false }
);

const reviewSchema = new Schema<IReview>(
  {
    userId: {
      type: Schema.Types.ObjectId,
      ref: "User",
      required: true,
      index: true,
    },
    title: {
      type: String,
      required: true,
      trim: true,
      maxlength: 150,
    },
    fileName: {
      type: String,
      required: true,
      trim: true,
      maxlength: 150,
    },
    codeType: {
      type: String,
      required: true,
      default: "typescript",
    },
    reviewMode: {
      type: String,
      required: true,
      default: "deep-review",
    },
    status: {
      type: String,
      enum: ["pending", "completed", "failed", "pending_approval", "approved", "rejected"],
      default: "completed",
      index: true,
    },
    originalCode: {
      type: String,
      required: true,
    },
    improvedCode: {
      type: String,
      default: "",
    },
    score: {
      type: Number,
      default: null,
      min: 0,
      max: 100,
    },
    summary: {
      type: String,
      default: "",
    },
    issues: {
      type: [issueSchema],
      default: [],
    },
    learningNotes: {
      type: [String],
      default: [],
    },
    recommendedTopics: {
      type: [String],
      default: [],
    },
    aiProvider: {
      type: String,
      default: "Gemini 2.5",
    },
    errorMessage: {
      type: String,
      default: "",
    },
    threadId: {
      type: String,
      default: "",
    },
  },
  {
    timestamps: true,
  }
);

reviewSchema.index({ userId: 1, createdAt: -1 });

export const ReviewModel: Model<IReview> =
  mongoose.models.Review || mongoose.model<IReview>("Review", reviewSchema);
